### PR TITLE
[Maps] Async Batch APIs for Forward/Reverse Geocode and Routing

### LINKED ISSUE #27158: [Azure Maps ] API Review
New API Review meeting has been requested.

**Service Name**: Azure Maps 
**Review Created By**: Danny Chin
**Review Date**: 02/01/2024 04:00 PM PT
**Onboarding Record**: 
**PR**: 
**Hero Scenarios Link**: Not Provided
**Core Concepts Doc Link**: Not Provided

**Description**: Async Batch APIs for forward geocode, reverse geocode, and routing

Detailed meeting information and documents provided can be accessed [here](https://web.powerapps.com/apps/e41c69de-ef96-44ba-b287-0e012e7c13ae?view=edit&event=7b44164b-a09f-ee11-be37-000d3a3418aa)
For more information that will help prepare you for this review, the requirements, and office hours, visit the documentation [here](https://aka.ms/azsdk/onboarding/restapischedule)

### PR DESCRIPTION
**Pull request by danielthank**
# Data Plane API - Pull Request

Add async batch APIs for forward geocode, reverse geocode, and routing.

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).

* Link to API Spec engagement record issue: https://github.com/Azure/azure-rest-api-specs/issues/27158

Is this review for (select one):

- [ ] a private preview
- [x] a public preview
- [ ] GA release

### Change Scope

<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous API Spec document (if applicable), and the root paths that have been updated. </sup>

* Design Document:
  * Service Design: https://microsoft.sharepoint.com/:w:/t/Azure-Maps/Eb8_fl1CuqpMkFYqGIW5cqUB5W-SY1MPbEwqOmS14mToVw?e=b3Bg37
  * API spec: https://microsoft.sharepoint.com/:w:/t/Azure-Maps/ERzkm1S1_ehPmrpXSKwtcogBruwDXaBQnKq99N9UeVa0-g?e=bwoPYY
* Previous API Spec Doc: NA
* Updated paths:

### Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

### Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/pr-suppressions) 
to get approval.

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling

 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://github.com/Azure/azure-api-style-guide/blob/main/README.md)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications

 * [Azure REST API Guidelines](https://aka.ms/azapi/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/azapi/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links

 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

<details>
  <summary>Checks stuck in `queued` state?</summary>
If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.  
</details>

### CODE REVIEW

**Comment by heaths** [comment]
I don't see a baseline commit, so it's hard to know exactly what's new in this api-version. I also see several RPs in this review. Do you have any documentation that highlights what is new for us to review? We likely will not have time to review the entirety of this PR.

**Review by heaths — CHANGES_REQUESTED** [review]
I left comments in the 3 APIViews I would like to see addressed.

**Comment by heaths** [comment]
Also, from the meeting, I see the confusion: you're treating each group of endpoints as a separate resource provider, which is atypical and confusing. You can put things in separate swaggers, but you should combine all swaggers as a single service in the top-level `readme.md` and ship together e.g., <https://github.com/Azure/azure-rest-api-specs/blob/main/specification/keyvault/data-plane/readme.md#tag-package-75>

**Comment by danielthank** [comment]
Justification of breaking changes

- Required status change from false to true
  New: [Common/stable/2023-06-01/common.json#L594:7](https://github.com/Azure/azure-rest-api-specs/blob/95ae2d0770cccd76f3b9bdd5da5cb9157f7bc90e/specification/maps/data-plane/Common/stable/2023-06-01/common.json#L594:7)
  Old: [common-types/data-plane/v1/types.json#L52:7](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/data-plane/v1/types.json#L52:7)
  Following Azure API guideline and it won't cause breaking change because we have already been returning the error response that is aligned with the Azure API guideline
- The new version has a different 'pattern' value than the previous one. (I added pattern constraint in `api-version`)

**Comment by weidongxu-microsoft** [comment]
> Let me try explain the problem of api-version.
> 
> ### case 1
> User initiates with REST API `/route/directions:asyncBatch?api-version=A`. During polling, SDK would invoke `/asyncBatch/operations/{operationId}?api-version=A` (even if your `operation-location` header from `/route` returns `api-version=B`)
> 
> This means "AsyncBatchManagement" Swagger would support all api-versions from any of the service doing the async.
> 
> ### case 2
> For now, your service does not have a resource model. So this does not apply. But when you have one, the reverse of case 1 would also happen.
> 
> User can start with REST API `/asyncBatch/operations/{operationId}?api-version=A`, which in the end (after LRO completed), could result in a resource in the other service e.g. `/route/someresource?api-version=A`

Loop in @JeffreyRichter for api-version discussion.

**Comment by danielthank** [comment]
@weidongxu-microsoft Thank you for the review.
Following the [POST or DELETE LRO pattern guideline](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#post-or-delete-lro-pattern), we will

1. Return the `Operation-Location` header with `api-version` the same as what users pass for job submission.
2. The status monitor endpoint will be backward compatible, which means the client can replace the `api-version` in `Operation-Location` header with newer api-version and get the status and result successfully.

For case 2, considering we have already decided to use POST pattern, we won't get a resource model in the status monitor. It means SDK users will not pass any part of the result of status monitor to another function in the SDK.

**Comment by weidongxu-microsoft** [comment]
For case 2, the model which contains "outputBlobUrl" is kind of a versioned response already (just a very simple one, for now).

You can have v1 with only "outputBlobUrl" in e.g. route; and v2 with e.g. "outputBlobUrl" and "additionalProperty". Whether to return "additionalProperty" in the response would depends on the `api-version` that initiate the `/asyncBatch/operations/{operationId}`.

And the service may grow. Today the route only have POST; it does not mean it is will always only have POST.

**Comment by danielthank** [comment]
@weidongxu-microsoft that's right. We have a model called `AsyncBatchOperationResult` in the swagger. There's nothing stopping us from extending the model in the future.
image

**Comment by danielthank** [comment]
@heaths @weidongxu-microsoft @JeffreyRichter We are ready to merge this PR. Please let us know if there's anything we need to change. Thank you.

**Comment by weidongxu-microsoft** [comment]
Please fix SpellCheck CI.

Also review BreakingChange CI, verify they are expected or false alert. If expected, you need approval from breaking change board.

**Comment by danielthank** [comment]
@weidongxu-microsoft I have written justification of SpellCheck and BreakingChange in previous discussion.

**Comment by JeffreyRichter** [comment]
You must justify the breaking changes: https://github.com/Azure/azure-rest-api-specs/pull/27188/checks?check_run_id=21209757976

The entire service must version uniformly. Please update all swaggers so that they all share the same api-version value.

**Comment by weidongxu-microsoft** [comment]
For spellcheck, click the "how to fix" link.
which points to https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/ci-fix.md#spell-check

Basically, update cspell.json

**Comment by chcmsft** [comment]
@pbrasil 
Please help to clarify the uniform api version concern raised by Jeffrey.

cc @jsedlak-microsoft

**Comment by chcmsft** [comment]
@heaths @weidongxu-microsoft 
The breaking change pipeline fail is fixed.
Can you help to review again and sign off if there are no further issue.

**Comment by chcmsft** [comment]
New change after the last review

- field linkedResource is replaced by outputStorageUrl. outputStorageUrl is the storage account blob storage url for storing the output result.

File: specification/maps/data-plane/Route/preview/2024-04-01-preview/route.json
Comment (weidongxu-microsoft) [inline specification/maps/data-plane/Route/preview/2024-04-01-preview/route.json:252]: Do you have a final schema after the LRO (e.g. which container in the storage account contains the results)?

If yes, please add `final-state-schema` refer to it.
https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-long-running-operation-options

File: specification/maps/data-plane/Route/preview/2024-04-01-preview/route.json
Comment (weidongxu-microsoft) [inline specification/maps/data-plane/Route/preview/2024-04-01-preview/route.json:2085]: Is this a storage account endpoint? Does your service access it and create the container and blobs?

Comment (chcmsft) [reply specification/maps/data-plane/Route/preview/2024-04-01-preview/route.json:2085]: It's a storage account endpoint. e.g. `https://uditest.blob.core.windows.net`

File: specification/maps/data-plane/AsyncBatchManagement/preview/2024-04-01-preview/asyncBatchManagement.json
Comment (weidongxu-microsoft) [inline specification/maps/data-plane/AsyncBatchManagement/preview/2024-04-01-preview/asyncBatchManagement.json:245]: description is wrong

File: specification/maps/data-plane/Route/preview/2024-04-01-preview/route.json
Comment (weidongxu-microsoft) [reply specification/maps/data-plane/Route/preview/2024-04-01-preview/route.json:2085]: I am wondering how user know which file in that `azure-maps-async-batch` container is the result of this e.g. route, especially when user do this async multiple times / for multiple routes, and give the same storage account endpoint `outputStorageUrl`.

Comment (chcmsft) [reply specification/maps/data-plane/Route/preview/2024-04-01-preview/route.json:2085]: In the Get Operation API, it has a field, `outputBlobUrl`, which points to the blob of the final result.

File: specification/maps/data-plane/AsyncBatchManagement/preview/2024-04-01-preview/asyncBatchManagement.json
Comment (chcmsft) [reply specification/maps/data-plane/AsyncBatchManagement/preview/2024-04-01-preview/asyncBatchManagement.json]: fixed

File: specification/maps/data-plane/Route/preview/2024-04-01-preview/route.json
Comment (chcmsft) [reply specification/maps/data-plane/Route/preview/2024-04-01-preview/route.json:256]: added

**Comment by chcmsft** [comment]
@weidongxu-microsoft all issues are addressed.

**Comment by chcmsft** [comment]
@heaths I've addressed all the comments above.
Do you have other concern?

File: specification/maps/data-plane/Route/preview/2024-04-01-preview/route.json
Comment (weidongxu-microsoft) [reply specification/maps/data-plane/Route/preview/2024-04-01-preview/route.json:2089]: Is there any design consideration that we have `outputStorageUrl` points to a storage account endpoint, instead of a e.g. `outputStorageBlobUrl` points to a single blob?

From the response and `outputBlobUrl`, it appears only output to 1 file. So, it seems natural for the API to let user specify the output file/blob location.

Comment (chcmsft) [reply specification/maps/data-plane/Route/preview/2024-04-01-preview/route.json:2089]: There are additional intermediate files that will also reside in the outputStorageUrl. Given our service's Bring Your Own Storage (BYOS) approach, we ensure that all files generated from customer input are stored in the storage provided by the customer, thereby allowing them full control over their data.

**Review by weidongxu-microsoft — APPROVED** [review]
LGTM for preview.

Versioning is the thing to check before GA https://github.com/Azure/azure-rest-api-specs/pull/27188#issuecomment-2041938136

**Comment by chcmsft** [comment]
@jhendrixMSFT Can you help to merge? I've got weidong's approval

**Comment by danielthank** [comment]
@jhendrixMSFT Could you help merge this PR? We have got review board's approval.

**Comment by jhendrixMSFT** [comment]
@JeffreyRichter has this been reviewed and approved? I don't see the necessary tags.

**Comment by JeffreyRichter** [comment]
@heaths was commenting in the APIView. Heath, are you ready to sign off on this PR?

**Comment by heaths** [comment]
@JeffreyRichter I don't have time to re-review so if you guys are happy with it, go ahead and merge. I don't want to sign off since I didn't review, but I can't seem to dismiss my review. The top doesn't show I'm blocking:

image

And there's no dismiss option in the summary at the bottom like I've seen on other PRs:

image

The author can't dismiss it either, which I know is possible because people have dismissed mine before (for good or bad reasons).

Can you dismiss it? Otherwise, I can just click the "Approve" option from the last screenshot above with a note I didn't review it.

**Review by heaths — APPROVED** [review]

**Comment by heaths** [comment]
You'll still need to update the branch before you can merge.

**Comment by weidongxu-microsoft** [comment]
"The entire service must version uniformly" must be solved before GA.

**Pull request merged**
